rust/private/rustfmt.bzl

```diff
@@ -0,0 +1,215 @@
+        tools = [toolchain.rustfmt],
+        arguments = [args],
+        mnemonic = "Rustfmt",
+    )
+
+    return [
+        OutputGroupInfo(
+            rustfmt = depset([marker]),
+        ),
+    ]
+
+rustfmt_check_aspect = aspect(
+    implementation = _rustfmt_check_aspect_impl,
+    fragments = ["cpp"],
+    host_fragments = ["cpp"],
+    toolchains = [
+        str(Label("//rust:toolchain")),
+    ],
+    attrs = {
+        "_process_wrapper": attr.label(
+            doc = "A process wrapper for running clippy on all platforms",
+            cfg = "exec",
+            executable = True,
+            default = Label("//util/process_wrapper"),
+        ),
+    },
+    incompatible_use_toolchain_transition = True,
+    doc = """\
+Executes rustfmt in `--check` mode on the specified target.
+
+To enable this aspect for your workspace, simply add the following to the `.bazelrc`
+file in the root of any workspace which loads `rules_rust`.
+
+```
+build --aspects=@rules_rust//rust:defs.bzl%rustfmt_check_aspect
+build --output_groups=+rustfmt
+```
+
+This aspect is executed on any target which provides the `CrateInfo` provider. However
+users may tag a target with `norustfmt` to have it skipped. Additionally, generated
+source files are also ignored by this aspect.
+""",
+)
+
+def _rustfmt_check_impl(ctx):
+    files = depset([], transitive = [target[OutputGroupInfo].rustfmt for target in ctx.attr.targets])
+    return [DefaultInfo(files = files)]
+
+rustfmt_check = rule(
+    implementation = _rustfmt_check_impl,
+    attrs = {
+        "targets": attr.label_list(
+            doc = "Rust targets to run rustfmt on.",
+            providers = [rust_common.crate_info],
+            aspects = [rustfmt_check_aspect],
+        ),
+    },
+    doc = """\
+A rule for defining a target which runs `rustfmt` in `--check` mode on an explicit list of targets
+
+For more information on the use of `rustfmt` directly, see [rustfmt_check_aspect](#rustfmt_check_aspect).
+""",
+)
+
+def rustfmt(name, config = Label("//tools/rustfmt:rustfmt.toml")):
+    """A macro defining a [rustfmt](https://github.com/rust-lang/rustfmt#readme) runner.
+
+    This macro is used to generate a rustfmt binary which can be run to format the Rust source
+    files of `rules_rust` targets in the workspace. To define this target, simply load and call
+    it in a BUILD file.
+
+    eg: `//:BUILD.bazel`
+
+    ```python
+    load("@rules_rust//rust:defs.bzl", "rustfmt")
+
+    rustfmt(
+        name = "rustfmt",
+    )
+    ```
+
+    This now allows users to run `bazel run //:rustfmt` to format any target which provides `CrateInfo`.
+
+    This binary also supports accepts a [label](https://docs.bazel.build/versions/master/build-ref.html#labels) or
+    pattern (`//my/package/...`) to allow for more granular control over what targets get formatted. This
+    can be useful when dealing with larger projects as `rustfmt` can only be run on a target which successfully
```

Comment:
Can aspects run on a target without also requiring that target to be built? For the `rustfmt` binaries to work, it needs a manifest of edition and sources. Can that be gathered without building the targets?

Reply:
Yeah, edition and sources are all statically knowable as part of analysis, and don't require any execution. I think if you remove the `+` from your `--output_groups` flag, it should do what you want without building

Reply:
>     1. I'm not sure why there are two near-identical aspects - could we drive the check vs fix based on which output group you're requesting, rather than by specifying different aspects? One aspect could register both actions, but only the ones driven by the selected `OutputGroup` would actually run.

I didn't realize I could use the output groups to control what actions were run. I thought it was an all or nothing process. It seems that by having creating `rustfmt_manifest` and `rustfmt_checks` output groups, I'm able to run the formatter without it first failing saying my code is not formatted. This is exactly what I want.

Reply:
I've updated this to remove `+` so you can format files without requiring that they get built and as a result removed `manual`.